### PR TITLE
feat: add subscription controller

### LINKED ISSUE #8: Develop CRD Subscription

### PR DESCRIPTION
**Pull request by Abirdcfly**
Fix #8

### CODE REVIEW

File: api/v1alpha1/subscription_types.go [hunk 1]
```diff
@@ -25,27 +25,29 @@ type InstallMethod string
 
 const (
 	// InstallMethodAuto means install directly without confirmation after detecting a new version.
-	InstallMethodAuto InstallMethod = "Auto"
+	InstallMethodAuto InstallMethod = "auto"
```
[the remark this answers is not in the document]
Comment (Abirdcfly) [reply api/v1alpha1/subscription_types.go:28]:
```go
// IsAuto returns true if InstallMethod is Auto, case-insensitivity.
func (i *InstallMethod) IsAuto() bool {
	return strings.EqualFold(string(*i), string(InstallMethodAuto))
}
```
Use the above function to judge the field. 

I am fine with these fields being in either uppercase or lowercase.🤷

No code style here...😳

File: controllers/subscription_controller.go [hunk 2]
```diff
@@ -18,45 +18,378 @@ package controllers
...
+	if err = r.Get(ctx, types.NamespacedName{Namespace: sub.Spec.RepositoryRef.Namespace, Name: sub.Spec.RepositoryRef.Name}, repo); err != nil {
+		logger.Error(err, "Failed to get repository", "Repository.Namespace", sub.Spec.RepositoryRef.Namespace, "Repository.Name", sub.Spec.RepositoryRef.Name)
+		if err1 := r.PatchCondition(ctx, sub, corev1alpha1.SubscriptionReconcileError(corev1alpha1.SubscriptionTypePlanSynce, err)); err1 != nil {
+			return errors.Wrap(err, err1.Error())
+		}
+		return err
+	}
+	// FIXME check if the repository is healthy
+	healthy := true
+	if repo.Status.GetCondition(corev1alpha1.TypeReady).Status != corev1.ConditionTrue {
```
Comment (bjwswang) [inline controllers/subscription_controller.go:318]: @0xff-dev do we have health check in repository? if yes, double check on this part.

**Review by bjwswang — APPROVED** [review]

File: api/v1alpha1/subscription_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bjwswang) [reply api/v1alpha1/subscription_types.go:28]: 🙆 👍 💯

File: controllers/subscription_controller.go (hunk 2 above)
Comment (0xff-dev) [reply controllers/subscription_controller.go:318]: No.

Comment (bjwswang) [reply controllers/subscription_controller.go:318]: 🆗

File: controllers/subscription_controller.go [hunk 3]
```diff
@@ -18,45 +18,378 @@ package controllers
...
+func (r *SubscriptionReconciler) PatchCondition(ctx context.Context, sub *corev1alpha1.Subscription, condition corev1alpha1.Condition) (err error) {
+	newSub := sub.DeepCopy()
+	newSub.Status.SetConditions(condition)
+	ready := true
+	for _, cond := range newSub.Status.Conditions {
+		if cond.Type == corev1alpha1.TypeReady {
+			continue
+		}
+		if cond.Status != corev1.ConditionTrue {
+			ready = false
```
Comment (0xff-dev) [inline controllers/subscription_controller.go:386]: If ready is false,  maybe the loop should just stop, there is no other case to change ready to true.

Comment (Abirdcfly) [reply controllers/subscription_controller.go:386]: There are 3 condition type: 
- Ready (final one)
- SourceSynced (indicates that the subscription is synced from component and repository )
- PlanSynced (indicates that the componentplan is synced from subscription)

I think "Ready" is the final result, and users and other controllers can use this condition type to determine whether the subscription is working properly.
SourceSynced and PlanSynced only explain in detail whether the internal two stages are normal.

Comment (0xff-dev) [reply controllers/subscription_controller.go:385]: I understand what you mean. What I mean is, add a break statement. There is no other case to set ready to true. It is possible to end the for loop early

**Review by 0xff-dev — APPROVED** [review]

**Pull request merged**